doc(framework): pytorch guide (#2735)
bentoml/BentoML: docs/source/frameworks/pytorch.rst

PyTorch
=======

> BentoML provides native support for serving and deploying models trained from PyTorch.

Reviewer: Add lowest version supported is 1.3.0

Preface
-------

User may also convert PyTorch models to torchscript models, and use ``bentoml.torchscript`` 
to save them. It will be a more stable way to serve and deploy PyTorch models in production.

.. note::

>     :bdg-info:`Remarks:` We recommend users apply model optimization techniques such as **distillation** or **quantization**.
>     Alternatively, PyTorch models can also be converted to :ref:`ONNX <frameworks/onnx>` models and leverage different runtimes (e.g. TensorRT, Apache TVM, etc.).

Reviewer: ```suggestion
    :bdg-info:`Remarks:` We recommend users to apply model optimization techniques such as `distillation <https://arxiv.org/abs/1503.02531>`_ or `quantization <https://pytorch.org/docs/stable/quantization.html#general-quantization-flow>`_ .
    Alternatively, PyTorch models can also be converted to :doc:`/frameworks/onnx` models and leverage different runtimes (e.g. TensorRT, Apache TVM, etc.).
```

Reviewer: add something like:
`in this tutorial, we will go over how to train a simple cnn with pytorch and use bentoml ...`

Saving a Trained Model
----------------------

For common PyTorch models with single input:

.. code-block:: python
    :caption: `train.py`

    import torch
    import torch.nn as nn
    import torch.nn.functional as F
    import torch.optim as optim

    class Net(nn.Module):
        def __init__(self):
            super().__init__()
            self.conv1 = nn.Conv2d(3, 6, 5)
            self.pool = nn.MaxPool2d(2, 2)
            self.conv2 = nn.Conv2d(6, 16, 5)
            self.fc1 = nn.Linear(16 * 5 * 5, 120)
            self.fc2 = nn.Linear(120, 84)
            self.fc3 = nn.Linear(84, 10)

        def forward(self, x):
            x = self.pool(F.relu(self.conv1(x)))
            x = self.pool(F.relu(self.conv2(x)))
            x = x.view(-1, 16 * 5 * 5)
            x = F.relu(self.fc1(x))
            x = F.relu(self.fc2(x))
            x = self.fc3(x)
            return x

    model = Net()

>     ... # training

Reviewer: add reproducible training code here.

also add a link to the prior tutorial if applicable.

    # =========

    bentoml.pytorch.save(model, "my_torch_model")

``bentoml.pytorch`` also supports saving models that take multiple tensors as input:

.. code-block:: python
    :caption: `train.py`

    import torch
    import torch.nn as nn
    import torch.nn.functional as F
    import torch.optim as optim

    class Net(nn.Module):
        def __init__(self):
            super().__init__()

        def forward(self, x, y):

>             return x + y

Reviewer: makes this into two sections with `1, 2` marking.

    model = Net()

>     ... # training

Reviewer: same here

    bentoml.pytorch.save(model, "my_torch_model")

.. note::

    :bdg-info:`Remarks:` All of the external python modules/util functions that the model
    requires must be included in the project. The PyTorch framework requires them to restore

>     the model.

Reviewer: makes this a warning. 

Also add an example how people should add the modules / utility functions to bento via `bentofile.yaml`.

You can use the YOLO example here.

.. note::

>     :code:`bentoml.pytorcpytorch.save_model` has parameter ``signatures``.

Reviewer: ```suggestion
    :code:`bentoml.pytorch.save_model` has parameter ``signatures``.
```

            loss.backward()
            optimizer.step()

            # print statistics
            print('Epoch: %d, Step: %d, Loss: %.4f' % (epoch, i, loss.item()))

    bentoml.pytorch.save(
        model,
        "my_torch_model",
        signatures={"__call__": {"batchable": True, "batchdim": 0}},
    )

``bentoml.pytorch`` also supports saving models that take multiple tensors as input:

.. code-block:: python
    :caption: `train.py`

    import torch
    import torch.nn as nn
    import torch.nn.functional as F
    import torch.optim as optim

    class Net(nn.Module):
        def __init__(self):
            super().__init__()

        def forward(self, x, y):
            return x + y

    model = Net()
    ... # training

    bentoml.pytorch.save(
        model,
        "my_torch_model",
        signatures={"__call__": {"batchable": True, "batchdim": 0}},
    )

.. note::

    :bdg-info:`Remarks:` All of the external python modules/util functions that the model
    requires must be included in the project. The PyTorch framework requires them to restore
    the model.

.. note::

    :code:`bentoml.pytorch.save_model` has parameter ``signatures``.

>     The ``signatures`` argument of type :ref:`Model Signatures <concepts/model:Model Signatures>`_ in :code:`bentoml.pytorch.save_model` is used to express which methods should be used for inference and exposed in the Runner. This signatures dictionary will be used during the creation process of a Runner instance.

Reviewer: ```suggestion
    The ``signatures`` argument of type :ref:`Model Signatures <concepts/model:Model Signatures>` in :obj:`bentoml.pytorch.save_model` is used to determine which methods will be used for inference and exposed in the Runner. The signatures dictionary will then be used during the creation process of a Runner instance.
```

    runner = bentoml.pytorch.get("my_torch_model").to_runner()

    svc = bentoml.Service(name="test_service", runners=[runner])

    @svc.api(input=JSON(), output=JSON())
    async def predict(json_obj: JSONSerializable) -> JSONSerializable:
        batch_ret = await runner.async_run([json_obj])
        return batch_ret[0]

.. note::

    Follow the steps to get the best performance out of your PyTorch model.
    #. Apply adaptive batching if possible.
    #. Serve on GPUs if applicable.
    #. See performance guide from `PyTorch Model Opt Doc <https://pytorch.org/tutorials/beginner/profiler.html>`_

Adaptive Batching
-----------------

Most PyTorch models can accept batched data as input. If batched interence is supported, it is recommended to enable batching to take advantage of 
the adaptive batching capability to improve the throughput and efficiency of the model. Enable adaptive batching by overriding the :code:`signatures` 
argument with the method name and providing :code:`batchable` and :code:`batch_dim` configurations when saving the model to the model store.

.. seealso::

   See :ref:`Adaptive Batching <guides/batching:Adaptive Batching>` to learn more.

.. note::

>    You can find more examples for **PyTorch** in our `bentoml/examples https://github.com/bentoml/BentoML/tree/main/examples`_ directory.

Reviewer: ```suggestion
   You can find more examples for **PyTorch** in our `bentoml/examples <https://github.com/bentoml/BentoML/tree/main/examples>`_ directory.
```

PyTorch
=======

BentoML provides native support for serving and deploying models trained from PyTorch. For more in-depth tutorials about PyTorch, please visit `PyTorch's official documentation <https://pytorch.org/tutorials/>`_

Preface
-------

> If you have already compiled your PyTorch model to TorchScript, you might consider to use :doc:`bentoml.torchscript </reference/frameworks/torchscript>`. BentoML provides first-class support for TorchScript, and hence using ``bentoml.torchscript`` are less prone to bugs during production.

Reviewer 2: ```suggestion
If you have already compiled your PyTorch model to TorchScript, you might consider to use :doc:`bentoml.torchscript </reference/frameworks/torchscript>`. BentoML provides first-class support for TorchScript, hence using ``bentoml.torchscript`` is less prone to compatibility issues during production.
```

.. note::

>     :bdg-info:`Remarks:` We recommend users to apply model optimization techniques such as `distillation <https://arxiv.org/abs/1503.02531>`_ or `quantization <https://pytorch.org/docs/stable/quantization.html#general-quantization-flow>`_ .
>     Alternatively, PyTorch models can also be converted to :doc:`/frameworks/onnx` models and leverage different runtimes (e.g. TensorRT, Apache TVM, etc.).

Reviewer 2: ```suggestion
    :bdg-info:`Remarks:` We recommend users to apply model optimization techniques such as `distillation <https://arxiv.org/abs/1503.02531>`_ or `quantization <https://pytorch.org/docs/stable/quantization.html#general-quantization-flow>`_ . Alternatively, PyTorch models can also be converted to :doc:`/frameworks/onnx` models and leverage different runtimes (e.g. TensorRT, Apache TVM, etc.) for better performance.
```

    trainloader = torch.utils.data.DataLoader(trainset, batch_size=batch_size,
                                              shuffle=True, num_workers=2)

    testset = torchvision.datasets.CIFAR10(root='./data', train=False,
                                           download=True, transform=transform)
    testloader = torch.utils.data.DataLoader(testset, batch_size=batch_size,
                                             shuffle=False, num_workers=2)

    classes = ('plane', 'car', 'bird', 'cat',
               'deer', 'dog', 'frog', 'horse', 'ship', 'truck')

    class Net(nn.Module):
        def __init__(self):
            super().__init__()
            self.conv1 = nn.Conv2d(3, 6, 5)
            self.pool = nn.MaxPool2d(2, 2)
            self.conv2 = nn.Conv2d(6, 16, 5)
            self.fc1 = nn.Linear(16 * 5 * 5, 120)
            self.fc2 = nn.Linear(120, 84)
            self.fc3 = nn.Linear(84, 10)

        def forward(self, x):
            x = self.pool(F.relu(self.conv1(x)))
            x = self.pool(F.relu(self.conv2(x)))
            x = x.view(-1, 16 * 5 * 5)
            x = F.relu(self.fc1(x))
            x = F.relu(self.fc2(x))
            x = self.fc3(x)

>             return x

Reviewer: IMO we should mimic how people actually do this in production, where they often separate a model file to a separate file. 

We then can add a warning like:
```
You can then add the following files to `bentofile.yaml`: ...
```

    model = Net()

    import torch.optim as optim

    criterion = nn.CrossEntropyLoss()
    optimizer = optim.SGD(net.parameters(), lr=0.001, momentum=0.9)

>     for epoch in range(2):

Reviewer 2: Let's add a comment stating the small epoch is just for demonstration purpose.

            loss.backward()
            optimizer.step()

            # print statistics
            print('Epoch: %d, Step: %d, Loss: %.4f' % (epoch, i, loss.item()))

    bentoml.pytorch.save(
        model,
        "my_torch_model",
        signatures={"__call__": {"batchable": True, "batchdim": 0}},
    )

``bentoml.pytorch`` also supports saving models that take multiple tensors as input:

.. code-block:: python
    :caption: `train.py`

    import torch
    import torch.nn as nn
    import torch.nn.functional as F
    import torch.optim as optim

    class Net(nn.Module):
        def __init__(self):
            super().__init__()

        def forward(self, x, y):
            return x + y

    model = Net()
    ... # training

    bentoml.pytorch.save(
        model,
        "my_torch_model",
        signatures={"__call__": {"batchable": True, "batchdim": 0}},
    )

.. note::

    :bdg-info:`Remarks:` All of the external python modules/util functions that the model
    requires must be included in the project. The PyTorch framework requires them to restore
    the model.

.. note::

    :code:`bentoml.pytorch.save_model` has parameter ``signatures``.

>     ``signatures`` refers to a general :ref:`Model Signatures <concepts/model:Model Signatures>`_ that dictates which methods can be used for inference in the Runner context. This signatures dictionary will be used during the creation process of a Runner instance.

Reviewer 2: ```suggestion
    The ``signatures`` argument of type :ref:`Model Signatures <concepts/model:Model Signatures>`_ in :code:`bentoml.pytorch.save_model` is used to express which methods should be used for inference and exposed in the Runner. This signatures dictionary will be used during the creation process of a Runner instance.
```

.. code-block:: python

    bentoml.pytorch.save(model, "my_model", signatures={"__call__": {"batch_dim": 0, "batchable": True}})

Building a Service
------------------

Create a BentoML service with the previously saved `my_torch_model` pipeline using the :code:`bentoml.pytorch` framework APIs.

.. code-block:: python
    :caption: `service.py`

    runner = bentoml.pytorch.get("my_torch_model").to_runner()

    svc = bentoml.Service(name="test_service", runners=[runner])

    @svc.api(input=JSON(), output=JSON())
    async def predict(json_obj: JSONSerializable) -> JSONSerializable:
        batch_ret = await runner.async_run([json_obj])
        return batch_ret[0]

.. note::

    Follow the steps to get the best performance out of your PyTorch model.
    #. Apply adaptive batching if possible.
    #. Serve on GPUs if applicable.

>     #. See performance guide from [PyTorch Doc]

Reviewer 2: Please link the docment.
